src/backend/storage/ipc/temp_tables_limit.c

```diff
@@ -0,0 +1,286 @@
+	{
+		Assert(prevFileLen >= 0);
+
+		newTotal = FileDiskSize(bufferedAppend->file) - prevFileLen;
+		Assert(newTotal == bytes); // for testing purposes
+		pg_atomic_add_fetch_u64(temp_tables_limit_value, newTotal);
+	}
+
+	bytes = 0;
+	prevFileLen = -1;
+}
+
+void
+TruncateAOSegmentFilePreHook(Relation rel, File fd)
+{
+	TempTablesLimitChecks();
+
+	if (rel->rd_islocaltemp)
+	{
+		prevFileLen = FileDiskSize(fd);
+	}
+}
+
+void
+TruncateAOSegmentFilePostHook(Relation rel, int64 offset)
+{
+	int64 bytesToTruncate;
+
+	TempTablesLimitChecks();
+
+	if (rel->rd_islocaltemp)
+	{
+		Assert(prevFileLen >= 0);
+
+		bytesToTruncate = prevFileLen - offset;
+		pg_atomic_sub_fetch_u64(temp_tables_limit_value, bytesToTruncate);
+	}
+
+	prevFileLen = -1;
+}
+
+/* heap */
+
+void
+mdextend_pre_hook(SMgrRelation reln, ForkNumber forknum, int64 size)
+{
+	TempTablesLimitChecks();
+
+	if (SmgrIsTemp(reln))
+		if (temp_tables_limit_value->value + size > TempTablesLimitToBytes()
+			&& forknum == MAIN_FORKNUM) /* visibility map can be extended during vacuum */
+			elog(ERROR, "Temp tables quota exceeded");
+}
+
+void
+mdextend_post_hook(SMgrRelation reln, int64 size)
+{
+	TempTablesLimitChecks();
+
+	if (SmgrIsTemp(reln))
+		pg_atomic_add_fetch_u64(temp_tables_limit_value, size);
+}
+
+void
+mdunlinkfork_pre_hook(RelFileNodeBackend rnode, ForkNumber forkNum)
+{
+	struct stat buf;
+	char *path;
+	int fd;
+
+	TempTablesLimitChecks();
+
+	if (RelFileNodeBackendIsTemp(rnode))
+	{
+		path = relpath(rnode, forkNum);
+		fd = OpenTransientFile((char *) path, O_RDWR | PG_BINARY, 0);
+		if (fstat(fd, &buf) == 0)
+		{
+			prevFileLen = buf.st_size;
+			CloseTransientFile(fd);
+		}
+		else
+			fileSkip = true; // file didn't exist, skip it
+	}
+}
+
+void
+mdunlinkfork_post_hook(RelFileNodeBackend rnode)
+{
+	TempTablesLimitChecks();
+
+	if (RelFileNodeBackendIsTemp(rnode) && !fileSkip)
+	{
+		Assert(prevFileLen >= 0);
+
+		pg_atomic_sub_fetch_u64(temp_tables_limit_value, prevFileLen);
+	}
+
+	prevFileLen = -1;
+	fileSkip = false;
+}
+
+void
+mdunlinkforksegment_pre_hook(RelFileNodeBackend rnode, char *segpath)
+{
+	struct stat buf;
+	char fullPath[MAXPGPATH];
+
+	TempTablesLimitChecks();
+
+	sprintf(fullPath, "%s/%s", data_directory, segpath);
+
+	if (RelFileNodeBackendIsTemp(rnode))
+		if (stat(fullPath, &buf) == 0)
+			prevSegFileLen = buf.st_size;
+		else
+			segFileSkip = true;
+}
+
+void
+mdunlinkforksegment_post_hook(RelFileNodeBackend rnode)
+{
+	TempTablesLimitChecks();
+
+	if (RelFileNodeBackendIsTemp(rnode) && !segFileSkip)
+	{
+		Assert(prevSegFileLen >= 0);
+
+		pg_atomic_sub_fetch_u64(temp_tables_limit_value, prevSegFileLen);
+	}
+
+	prevSegFileLen = -1;
+	segFileSkip = false;
+}
+
+void
+mdunlink_ao_perFile_pre_hook(char *segPath)
+{
+	struct stat buf;
+	char *segPathCopy = malloc(sizeof(char) * strlen(segPath));
+	char fullPath[MAXPGPATH];
+	char *name;
+
+	TempTablesLimitChecks();
+
+	strcpy(segPathCopy, segPath);
```

Comment:
```suggestion
	segPathCopy = strdup(segPath);
```
and explicit malloc is not needed